src/hooks/filters.ts

```diff
@@ -0,0 +1,294 @@
+import { keys } from "@/constants";
+import type {
+  CheckboxItems,
+  CheckboxItemsByFilterName,
+  CheckboxState,
+  CheckedChangePayload,
+  CheckedFiltersByFilterName,
+  FilterName,
+  OracleQueryUI,
+} from "@/types";
+import Fuse from "fuse.js";
+import { cloneDeep } from "lodash";
+import { useEffect, useMemo, useReducer, useState } from "react";
+
+/**
+ * Combines the filter and search hooks
+ * @returns results - the filtered and searched queries
+ * @returns filterProps - the props for the filter checkboxes
+ * @returns searchProps - the props for the search input
+ */
+export function useFilterAndSearch(queries: OracleQueryUI[] | undefined = []) {
+  const { filteredQueries, ...filterProps } = useFilters(queries);
+  const { results, ...searchProps } = useSearch(filteredQueries);
+
+  return {
+    results,
+    filterProps,
+    searchProps,
+  };
+}
+
+type State = {
+  filters: CheckboxItemsByFilterName;
+  checkedFilters: CheckedFiltersByFilterName;
+  filteredQueries: OracleQueryUI[];
+};
+
+type Action = MakeEntries | CheckedChange | Reset;
+
+type MakeEntries = {
+  type: "make-entries";
+};
+
+type Reset = {
+  type: "reset";
+};
+
+type CheckedChange = {
+  type: "checked-change";
+  payload: CheckedChangePayload;
+};
+
+/**
+ * Filter the queries based on the checked filters
+ * @returns filters - state of the filter checkboxes
+ * @returns checkedFilters - the checked filters in an array for rendering a list of checked filters
+ * @returns filteredQueries - the filtered queries
+ * @returns checkedChange - update the filters based on the checked state of a checkbox
+ * @returns reset - reset the filters to their initial state
+ */
+export function useFilters(queries: OracleQueryUI[]) {
+  const initialState: State = {
+    filters: {
+      project: { All: { checked: true, count: 0 } },
+      chainName: { All: { checked: true, count: 0 } },
+      oracleType: { All: { checked: true, count: 0 } },
+    },
+    checkedFilters: {
+      project: [],
+      chainName: [],
+      oracleType: [],
+    },
+    filteredQueries: cloneDeep(queries),
+  };
+  const [state, dispatch] = useReducer(reducer, initialState);
+
+  useEffect(() => {
+    dispatch({ type: "make-entries" });
+  }, [queries]);
+
+  function reducer(state: State, action: Action) {
```

Comment:
not a huge deal, but it looks like this only relies on `queries` in the closure. Theoretically you could move this function out of the hook and use currying like `useReducer(Reducer(queries))` in the hook where `Reducer = queries => reducer`. not a huge deal but in practice its nicer if you can export and test this logic independent of the hook

Reply:
great feedback thank you. I've implemented this now, will request review soon